x-pack/plugins/ml/public/application/components/scatterplot_matrix/scatterplot_matrix.tsx

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useMemo, useEffect, useState, FC } from 'react';
+
+// There is still an issue with Vega Lite's typings with the strict mode Kibana is using.
+// @ts-ignore
+import { compile } from 'vega-lite/build-es5/vega-lite';
+import { parse, View, Warn } from 'vega';
+import { Handler } from 'vega-tooltip';
+
+import {
+  htmlIdGenerator,
+  EuiComboBox,
+  EuiComboBoxOptionOption,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiFormRow,
+  EuiLoadingSpinner,
+  EuiSelect,
+  EuiSpacer,
+  EuiSwitch,
+  EuiText,
+} from '@elastic/eui';
+
+import { i18n } from '@kbn/i18n';
+
+import type { SearchResponse7 } from '../../../../common/types/es_client';
+
+import { ml } from '../../services/ml_api_service';
+
+import { getProcessedFields } from '../data_grid';
+
+import {
+  getScatterplotMatrixVegaLiteSpec,
+  LegendType,
+  OUTLIER_SCORE_FIELD,
+} from './scatterplot_matrix_vega_lite_spec';
+
+import './scatterplot_matrix.scss';
+
+const SCATTERPLOT_MATRIX_DEFAULT_FIELDS = 4;
+const SCATTERPLOT_MATRIX_DEFAULT_FETCH_SIZE = 1000;
+const SCATTERPLOT_MATRIX_DEFAULT_FETCH_MIN_SIZE = 1;
+const SCATTERPLOT_MATRIX_DEFAULT_FETCH_MAX_SIZE = 10000;
+
+const TOGGLE_ON = i18n.translate('xpack.ml.splom.toggleOn', {
+  defaultMessage: 'On',
+});
+const TOGGLE_OFF = i18n.translate('xpack.ml.splom.toggleOff', {
+  defaultMessage: 'Off',
+});
+
+const sampleSizeOptions = [100, 1000, 10000].map((d) => ({ value: d, text: '' + d }));
+
+interface ScatterplotMatrixProps {
+  fields: string[];
+  index: string;
+  resultsField?: string;
+  color?: string;
+  legendType?: LegendType;
+}
+
+export const ScatterplotMatrix: FC<ScatterplotMatrixProps> = ({
+  fields: allFields,
+  index,
+  resultsField,
+  color,
+  legendType,
+}) => {
+  const defaultFields =
+    allFields.length > SCATTERPLOT_MATRIX_DEFAULT_FIELDS
+      ? allFields.slice(0, SCATTERPLOT_MATRIX_DEFAULT_FIELDS)
+      : allFields;
+
+  // dynamicSize is optionally used for outlier charts where the scatterplot marks
+  // are sized according to outlier_score
+  const [dynamicSize, setDynamicSize] = useState<boolean>(false);
+
+  // used to give the use the option to customize the fields used for the matrix axes
+  const [fields, setFields] = useState<string[]>(defaultFields);
```

Comment:
Would be nice to have these stored in the URL as well as the collapse state of the panel.

Reply:
Collapse state is store in the URL now, added an item about selected fields in the meta issue.